### ISSUE TITLE
`shadow_unrelated` raised when assigning couples

### ISSUE THREAD

**Issue by t-webber** [rust-lang/rust-clippy#14377]
### Summary

`shadow_unrelated` is fired when I assign a couple of variables, even if they are not shadowing anything; it is even fired if the variables were not assigned

### Lint Name

shadow_unrelated

### Reproducer

I tried this code:

```rust
#![allow(unused)]
#![warn(clippy::shadow_unrelated)]

fn main() {
    let a;
    let b;
    (a, b) = (0, 1);
}
```

I saw this happen:

```
warning: `b` shadows a previous, unrelated binding
 --> src/main.rs:7:9
  |
7 |     (a, b) = (0, 1);
  |         ^
  |
note: previous binding is here
 --> src/main.rs:7:6
  |
7 |     (a, b) = (0, 1);
  |      ^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#shadow_unrelated
note: the lint level is defined here
 --> src/main.rs:2:9
  |
2 | #![warn(clippy::shadow_unrelated)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^

```

I expected to see this happen: nothing

### Version

```text
rustc 1.87.0-nightly (f5a1ef712 2025-03-07)
binary: rustc
commit-hash: f5a1ef7121ad661b5a21a1d02941c8064d54ee0b
commit-date: 2025-03-07
host: x86_64-unknown-linux-gnu
release: 1.87.0-nightly
LLVM version: 20.1.0
```

**Comment by samueltardieu** [comment]
Fun problem, which is due to the way tuple assignment is transformed by the compiler:

```rust
fn main() { let a; let b; { let (lhs, lhs) = (0, 1); a = lhs; b = lhs; }; }
```

@rustbot claim

**Comment by profetia** [comment]
I wonder if this `AssignDesugar` behavior is newly introduced in the compiler. I also encountered one in #14371. Seems that it was not considered in clippy in the past.

**Comment by samueltardieu** [comment]
> I wonder if this `AssignDesugar` behavior is newly introduced in the compiler. I also encountered one in [#14371](https://github.com/rust-lang/rust-clippy/pull/14371). Seems that it was not considered in clippy in the past.

It has been introduced in 2020, but even the compiler documentation was incomplete until [last September](https://github.com/rust-lang/rust/commit/a47e9b6c).
